ci/test_suites/authenticators_oidc/test

```diff
@@ -0,0 +1,117 @@
+  )
+
+  # shellcheck disable=SC2034
+  arr=(
+    "${keycloak_items[@]}"
+    "PROVIDER_URI=https://keycloak:8443/auth/realms/master"
+    "PROVIDER_INTERNAL_URI=http://keycloak:8080/auth/realms/master/protocol/openid-connect"
+    "ID_TOKEN_USER_PROPERTY=preferred_username"
+  )
+}
+
+# The arguments must be unexpanded variable names.  Eg:
+#
+# _create_keycloak_user '$APP_USER' '$APP_PW' '$APP_EMAIL'
+#
+# This is because those variables are not available to this script. They are
+# available to bash commands run via "docker-compose exec oidc-keycloak bash
+# -c...", since they're defined in the docker-compose.yml.
+_create_keycloak_user() {
+  local user_var=$1
+  local pw_var=$2
+  local email_var=$3
+
+  docker-compose exec -T \
+    oidc-keycloak \
+    bash -c "/scripts/create_user \"$user_var\" \"$pw_var\" \"$email_var\""
+}
+
+# TODO: use wait_for_cmd
+_wait_for_keycloak() {
+  for i in {1..40}; do
+    sleep=5
+    echo "Keycloak starting logs:"
+
+    if docker-compose logs oidc-keycloak | grep "started"; then
+      echo "Keycloak server is up and ready"
+      return 0
+    fi
+
+    echo "Keycloak not ready yet sleep number $i for $sleep seconds"
+    sleep "$sleep"
+  done
+
+  echo "Error with keycloak server start or it is too slow"
+  exit 1
+}
+
+_prepare_env_auth_oidc() {
+  docker-compose up --no-deps -d pg conjur oidc-keycloak
+  _wait_for_keycloak
+
+  echo "Defining oidc-keycloak client"
+
+  docker-compose exec -T oidc-keycloak /scripts/create_client
+
+  echo "Creating user 'alice' in Keycloak"
+
+  # Note: We want to pass the bash command thru without expansion here.
+  # shellcheck disable=SC2016
+  _create_keycloak_user \
+    '$KEYCLOAK_APP_USER' \
+    '$KEYCLOAK_APP_USER_PASSWORD' \
+    '$KEYCLOAK_APP_USER_EMAIL'
+
+  echo "Creating second user 'bob' in Keycloak"
+
+  # Note: We want to pass the bash command thru without expansion here.
+  # shellcheck disable=SC2016
+  _create_keycloak_user \
+    '$KEYCLOAK_SECOND_APP_USER' \
+    '$KEYCLOAK_SECOND_APP_USER_PASSWORD' \
+    '$KEYCLOAK_SECOND_APP_USER_EMAIL'
+
+  echo "Creating user in Keycloak that will not exist in conjur"
+
+  # Note: We want to pass the bash command thru without expansion here.
+  # shellcheck disable=SC2016
+  _create_keycloak_user \
+    '$KEYCLOAK_NON_CONJUR_APP_USER' \
+    '$KEYCLOAK_NON_CONJUR_APP_USER_PASSWORD' \
+    '$KEYCLOAK_NON_CONJUR_APP_USER_EMAIL'
+
+  echo "Initialize keycloak certificate in conjur server"
+
+  # TODO: there's a dep on the docker-compose.yml volumes
+  docker-compose exec -T \
+    conjur /authn-oidc/keycloak/scripts/fetchCertificate
+}
+
+# We also run an ldap-server container for testing the OIDC & LDAP combined use-case.
+# Note: We can't run the above use-case in a separate Jenkins step because we'll have a port bind for keycloak
+_prepare_env_auth_ldap
+_prepare_env_auth_oidc
+
+_run_cucumber_tests authenticators_oidc 'ldap-server oidc-keycloak' \
+  _hydrate_oidc_env_args
```

Comment:
What are your thoughts on introducing a `main` function when you have other functions in the script? It's a personal preference, but I'd rather see the entrypoint flow at the top, with supporting functions beneath.

Reply:
I think it's a good pattern, I'll make that change.